### ISSUE TITLE
[ML] Metadata Migration Meta issue

### ISSUE THREAD

**Issue by davidkyle** [elastic/elasticsearch#32905]
Anomaly detector jobs and datafeed configurations are currently stored in the cluster state, the initial design decision was based on the need to distribute work across the cluster. After persistent tasks were invented -and with a little hindsight- it is not necessary to store config in the cluster state if the persistent task parameters contain enough information to open a job and start a datafeed. 

Placing config in the cluster state has caused a number of issues:

- Jobs cannot be easily transferred to another cluster
- User defined searches in the datafeed may use deprecated features removed in the next version. After upgrading this can fail reading the cluster state as parsing the removed feature throws an exception 
- If you restore a snapshotted cluster state you restore all jobs that existed at the time the snapshot was taken
- Jobs do not recover if the cluster state is lost #30088
- Adding new types of job or changing the validation is difficult or impossible in a backwards compatible manner #30084 #30070

### Proposal
Starting in 6.last new job and datafeed configurations will be stored in a new internal index `.ml-config`. Jobs created prior to 6.last will be migrated to index documents and removed from the cluster state with the goal of removing all ml data from the cluster state. 7.x will retain the ability to  read jobs from the cluster state to support full cluster upgrades and restoring a snapshot containing a global cluster state. The preferred solution is to automatically migrate extant cluster state jobs on allocation. In the 8 series all the code to handle ml config in cluster state will be dropped. 

### Work Plan
Work will be done on the feature branch `feature/feature-jindex-6x` and `feature/feature-jindex-master` with regular merges. The first stage is the changes required to create and run a job with its configuration stored in an index. Once that is stable and passing the testing gate the migration of existing jobs on upgrade will be tackled.  

### Phase 1: Run Jobs with their configuration defined in an index
- [x] Add the `.ml-config` index template with job and datafeed mappings #32719
- [x] Add class to handle read/write/update/deletes on config documents #32854 #32738
- [x] Wildcard expansion for getting jobs and datafeed e.g. `_all`, `foo-*`. This replaces the MlMetaData Group or Job Lookup functionality #32854 #32738
- [x] Change JobManager to operate on migrated configs #33064
- [x] Change DatafeedManager to operate on migrated configs #33273
- [x] The OpenJobPersistentTasksExecutor validate and selectLeastLoadedMlNode methods require information that can no longer be read from the cluster state #34084
- [x] DatafeedNodeSelector requires information that can no longer be read from the cluster state #34218
- [x] Close job and TransportFinalizeJobExecutionAction must be changed so it doesn't update the cluster state #34217 #34226
- [x] AbstractExpiredJobDataRemover must read migrated job configurations #34532
- [x] Delete job and implement the equivalent of markJobAsDeleting #34595
- [x] Get job and datafeed stats #34645 
- [x] All the endpoints #34605 #34642 #34710
- [x] All the tests (excluding rolling upgrade) #34851
- [x] Replace Version.CURRENT with the release version #36118

### Phase 1a: 6.6 & 6.7 Jobs can be defined in the clusterstate or an index document
- [x] All reads of config must check both index and clusterstate #35344 #35539 #35590 #35598
- [x] Upgrade tests #35700 #36593

### Phase 2: Migrate Job and Datafeed Configuration
- [x] Where config exists in both index and clusterstate prefer ~~index~~ clusterstate #35940 #36014
- [x] Create a migration class for jobs and datafeeds #35834
- [x] Automatically migrate closed jobs and datafeeds #35834 #36425 #36481
- [x] Migrate open jobs and datafeeds once they become unallocated #37430 #37536

### Issues
- [x] Alternative method to share job memory usage. **Blocker** #34084 #35263
- [x] Parsing deprecated and removed features in Datafeed Config **Blocker** #34858
- [x] Backup MlMetadata before starting migration #36422 
- [x] Repeatedly migrate config in batches #36716 
- [x] Create .`ml-config` index in case autocreate is disabled #36608
- [x] Transient setting override #36700
- [x] Check .ml-config is yellow #36608
- [x] CI failure MlDistributedFailureIT.testLoseDedicatedMasterNode #36760
- [x] CI failure RestoreModelSnapshotIT #36849
- [x] CI failure MlMigrationFullClusterRestartIT #36816
- [x] CI failure 60_ml_config_migration #36810
- [x] CI failure model mem limit #36961
- [x] NPE in MachineLearningFeatureSet.addJobsUsage elastic/ml-cpp#351
- [x] UnusedStateRemover will remove all state for jobs not in cluster state **Blocker** #37109
- [x] MachineLearningLicensingTests #37117 

### Nice to haves
- [ ] TransportJobTaskAction no longer throws missing job #34747
- [ ] Get Jobs response is limited by search size  #34864 #36608
- [ ] Consider changing where config is validated #34899 
- [ ] Use custom all field for configs #36445 
- [x] Don't write empty lists of jobs and datafeeds in MlMetadata.toXContent #36421
- [ ] Index aliases for .ml-config
- [ ] OpenJobParams (p task params) contain the full job when only certain fields are required
- [ ] Ensure ordering of updates to the autodetect process

**Comment by davidkyle** [comment]
pinging @elastic/ml-core

**Comment by davidkyle** [comment]
pinging @bleskes

**Comment by bleskes** [comment]
@davidkyle looks good. I miss the part mentioning that the job upgrader automatically picks up on unallocated old style job (and leaves allocated ones alone). Is that still part of the plan?

**Comment by droberts195** [comment]
@bleskes in 6.last that is definitely the plan.  The job upgrader will leave allocated jobs alone to allow for a seamless rolling upgrade from 6.previous.

Once the master is on 7.x I believe the plan is to aggressively upgrade all jobs, even if that means stopping them.  However, this should not really be a problem because you cannot do a rolling upgrade to 7.x until every node is on at least 6.last, and the process of doing a rolling upgrade to 6.last would have reallocated every job already.  And if somebody does a full cluster restart from 6.x to 7.x then every job will need allocating anyway.

@davidkyle is on holiday until Wednesday, so if you need more detail then he can provide it when he's back.

**Comment by dimitris-athanasiou** [comment]
Adding a reminder that once migration starts, non-migrated jobs should not be allowed to update in any way including opening.

**Comment by davidkyle** [comment]
This project was completed long ago, the issue can be closed now
